File: content/courses/parallel-computing-introduction/solns/mpi_nonblock_2dplate.f90
program sendrows
   use mpi_f08

   integer, parameter :: maxiter=10000000
   integer            :: N, M
   integer            :: i,j
   integer            :: numargs, diff_interval, interations=0
   integer            :: lb1, lb2
   double precision   :: eps, diff = 1.0
   character(len=80)  :: arg, filename
   character(len=4)   :: char_row, char_col
   double precision, allocatable, dimension(:,:)  :: u, w
   double precision   :: time0, time1

   double precision   :: rows
   integer            :: nrows, ncols, nrl, ncl
   integer            :: rank, nprocs, tag=0
   integer            :: left, right, up, down
   integer, dimension(2) :: dims, grid_coords
   logical, dimension(2) :: periods
   logical            :: reorder
   integer            :: ndims, grid_rank
   integer            :: direction, displ
   integer, parameter :: root=0
   integer            :: nrequests
   type(MPI_Request), dimension(:), allocatable :: mpi_requests
   type(MPI_Comm)     :: grid_comm
   type(MPI_Datatype) :: row
   double precision   :: gdiff
   character(len=24)  :: fname

   interface
     subroutine set_bcs(nr,nc,nrows,ncols,coords,u)
        implicit none
        integer,                intent(in)  :: nr, nc, nrows, ncols
        integer, dimension(:),  intent(in)                :: coords
        double precision, dimension(0:,0:), intent(inout) :: u
     end subroutine
   end interface

   !Initialize MPI, get the local number of columns
   call MPI_INIT()
   call MPI_COMM_SIZE(MPI_COMM_WORLD,nprocs)
   call MPI_COMM_RANK(MPI_COMM_WORLD,rank)

   ! check number of parameters and read in epsilon, filename, optionally size
   ! all ranks do this
   numargs=command_argument_count()
   if (numargs .lt. 2) then
      call MPI_Finalize()
      stop 'USAGE: epsilon output-file <N> <M>'
   else
      call get_command_argument(1,arg)
      read(arg,*) eps
      call get_command_argument(2,filename)

      if (numargs==2) then
         N=500
         M=500
      endif
      if (numargs==3) then
         call get_command_argument(3,arg)
         read(arg,*) N
         M=N
      endif
      if (numargs==4) then
         call get_command_argument(3,arg)
         read(arg,*) N
         call get_command_argument(4,arg)
         read(arg,*) M
      endif
   endif

   !For simplicity, we will limit this code to perfect square process count
   rows=sqrt(dble(nprocs))
   if ( rows /= dble(int(rows)) ) then
       call MPI_Finalize()
       stop "This code requires a perfect square number of processes"
   else
       nrows=int(rows)
       ncols=nrows
   endif

   !Strong scaling
   !Weak scaling would set nrl=N and ncl=M
   if (mod(N,nrows)==0 .and. mod(M,ncols)==0) then
       nrl = N/nrows
       ncl = M/ncols
   else
       call MPI_Finalize()
       stop "Number of ranks should divide the number of rows evenly."
   endif

   !create cartesian topology for processes
   ndims=2
   dims(1) = nrows      ! number of rows
   dims(2) = ncols      ! number of columns
   periods(1) = .false.  ! cyclic in this direction
   periods(2) = .false. ! not cyclic in this direction
   reorder    = .false.  ! allow MPI to reorder if it sees fit
   call MPI_Cart_create(MPI_COMM_WORLD, ndims, dims, periods,                 &
                                                     reorder, grid_comm)
   call MPI_Comm_rank(grid_comm, grid_rank)
   call MPI_Cart_coords(grid_comm, grid_rank, ndims, grid_coords)

   direction =  0   ! shift along the 1st index (0 or 1)
   displ =  1       ! shift by  1
   call MPI_Cart_shift(grid_comm, direction, displ, up, down)

   direction =  1   ! shift along the 2nd index (0 or 1)
   call MPI_Cart_shift(grid_comm, direction, displ, left, right)

   write(*,*) 'topo', grid_rank, up, down, left, right


   ! Set up values
   lb1=0
   lb2=0
   allocate(u(lb1:nrl+1, lb2:ncl+1), w(lb1:nrl+1,lb2:ncl+1))

   u=0.d0
   w=0.d0

   ! Boundary conditions
   call set_bcs(nrl,ncl,nrows,ncols,grid_coords,u)

   call MPI_Type_vector(ncl+2,1,nrl+2,MPI_DOUBLE_PRECISION,row)
   call MPI_TYPE_COMMIT(row)

   nrequests=8
   allocate(mpi_requests(nrequests))

   diff_interval=1

  ! Walltime from process 0 is good enough for me
   if (grid_rank .eq. 0) then
      write (*,'(a,es15.8,a,i4,a,i4)') 'Running until the difference is <= ', &
                                              eps,' for global size ',N,'x',M
     time0=MPI_WTIME()
   endif

  ! Compute steady-state solution
   do while ( iterations<=maxiter )

     ! Exchange halo values

   call MPI_Irecv(u(1:nrl,ncl+1),nrl,MPI_DOUBLE_PRECISION,right,tag,          &
                                     grid_comm,mpi_requests(1))
   call MPI_Irecv(u(1:nrl,0)    ,nrl,MPI_DOUBLE_PRECISION,left,tag,           &
                                     grid_comm,mpi_requests(2))
   call MPI_Irecv(u(0,0)     ,1,row,up,tag, grid_comm,mpi_requests(3))
   call MPI_Irecv(u(nrl+1,0) ,1,row,down,tag, grid_comm,mpi_requests(4))


   call MPI_Isend(u(1:nrl,1)    ,nrl,MPI_DOUBLE_PRECISION,left,tag,           &
                                     grid_comm,mpi_requests(5))
   call MPI_Isend(u(1:nrl,ncl)  ,nrl,MPI_DOUBLE_PRECISION,right,tag,          &
                                     grid_comm,mpi_requests(6))
   call MPI_Isend(u(nrl,0)  ,1,row,down,tag, grid_comm,mpi_requests(8))
   call MPI_Isend(u(1,0),    1,row,up,tag, grid_comm,mpi_requests(7))

       do j=1,ncl
           do i=1,nrl
               w(i,j) = 0.25*(u(i-1,j) + u(i+1,j) + u(i,j-1) + u(i,j+1))
           enddo
       enddo

   ! complete communications
   call MPI_Waitall(nrequests,mpi_requests,MPI_STATUSES_IGNORE)

       if (mod(iterations,diff_interval)==0) then
           if (diff_interval==-1) continue  !disable convergence test
               diff=maxval(abs(w(1:nrl,1:ncl)-u(1:nrl,1:ncl)))
               call MPI_ALLREDUCE(diff,gdiff,1,MPI_DOUBLE_PRECISION,MPI_MAX,  &
                                                             MPI_COMM_WORLD)

               if (gdiff <= eps) then
                   exit
               endif
       endif

      !Set halo values
       w(0,:)=u(0,:)
       w(nrl+1,:)=u(nrl+1,:)
       w(:,0)=u(:,0)
       w(:,ncl+1)=u(:,ncl+1)

       u = w

     ! Reset physical boundaries (they get overwritten in the halo exchange)
       call set_bcs(nrl,ncl,nrows,ncols,grid_coords,u)

      !If this happens we will exit at next conditional test.
      !Avoids explicit break here
       if (iterations>maxiter) then
          if (grid_rank==0) then
              write(*,*) "Warning: maximum iterations exceeded"
          endif
       endif

       iterations = iterations + 1

   enddo

   if (grid_rank==root) then
     time1 = MPI_WTIME()
     write(*,*) 'completed; iterations = ', iterations,"in ",time1-time0," sec"
   endif


  ! Write solution to output file by row. Must use correct stitching to get
  ! overall result, such as a numpy concatenate on axis=1.
  ! It would also work to output by columns, in which case stitching is more
  ! straightforward.

  !Some compilers (Intel) like to add EOLs to line up columns in output. The
  ! Fortran standard permits this with * IO, but it complicates plotting.
  ! Use some F08 features to fix this.

  write(char_row,'(i4)')  grid_coords(1) 
  write(char_col,'(i4)')  grid_coords(2) 
  fname=trim(filename)//trim(adjustl(char_row))//trim(adjustl(char_col))

  open (unit=10,file=fname)
  do i=1,nrl
     write (10,'(*(g0,1x))') u(i,1:ncl)
  enddo


   call MPI_Comm_free(grid_comm)

   call MPI_Finalize()

end program


subroutine set_bcs(nr,nc,nrows,ncols,coords,u)
  implicit none
  integer, intent(in)                      :: nr, nc, nrows, ncols
  integer, dimension(:), intent(in)        :: coords
  double precision, dimension(0:,0:), intent(inout) :: u
  double precision                         :: topBC, bottomBC, leftBC, rightBC

  ! Set boundary values
  ! This has the ice bath on the bottom edge.
  ! Note: when plotting, 0,0 is the bottom.

  topBC=0.d0
  bottomBC=100.d0
  rightBC=100.d0
  leftBC=100.d0

  !Set boundary conditions
   if (coords(1)==0) then
      u(0,:)=topBC
   else if (coords(1)==nrows-1) then
      u(nr+1,:)=bottomBC
   endif
   if (coords(2)==0) then
      u(:,0)=leftBC
   else if (coords(2)==ncols-1) then
      u(:,nc+1)=rightBC
   endif

end subroutine set_bcs

File: content/courses/parallel-computing-introduction/solns/mpi_nonblock_2dplate.cxx
#include <iostream> 
#include <cmath>
#include <string>
#include <sstream>
#include <fstream>
#include <iomanip> 
#include <mpi.h> 


#define MAX_ITER 10000000

void set_bcs(int nr, int nc, int nrows, int ncols, int* coords, double **u);

using namespace std; 

int main (int argc, char *argv[]) {

    double diff;            // Change in value
    int N, M;
    int iterations = 0;
    int diffInterval;
    double epsilon;

    // Added for MPI
    int nrows, ncols;
    int nrl, ncl;
    int rank, nprocs;
    int errcode;
    int up, down, right, left;
    int tag=0;
    double gdiff;

    //Initialize MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    /* check number of parameters and read in epsilon, filename, optionally size */
   if (argc < 3) {
     printf ("USAGE:  %s epsilon output-file <N> <M>\n", argv[0]);
     MPI_Abort(MPI_COMM_WORLD,errcode);
     return 1;
   }
   else {
      stringstream ssepsilon;
      ssepsilon<<argv[1];
      if (ssepsilon.fail()) {
         printf ("Error reading in epsilon value\n");
         MPI_Abort(MPI_COMM_WORLD,errcode);
         return 2;
      }
      ssepsilon>>epsilon;

      if (argc==3) {
        N=500;
        M=500;
      }
      if (argc==4) {
        stringstream ssnr;
        ssnr<<argv[3];
        if (ssnr.fail()) {
           cout<<"Error converting row dimension \n";
           MPI_Abort(MPI_COMM_WORLD,errcode);
           return 2;
         }
         ssnr>>N;
         M=N;
      }
      if (argc==5) {
        stringstream ssnr;
        ssnr<<argv[3];
        if (ssnr.fail()) {
            cout<<"Error converting row dimension \n";
            MPI_Abort(MPI_COMM_WORLD,errcode);
            return 2;
        }
        ssnr>>N;
        stringstream ssnc;
        ssnc<<argv[4];
        if (ssnc.fail()) {
            cout<<"Error converting row dimension \n";
            MPI_Abort(MPI_COMM_WORLD,errcode);
            return 2;
        }
        ssnc>>M;
     }
   }
   //For simplicity, we will limit this code to perfect square process count
   double rows=sqrt(nprocs);
   nrows=(int)rows;
   if ( nrows*nrows != nprocs ) {
       cout<<"This code requires a perfect square number of processes\n";
       MPI_Abort(MPI_COMM_WORLD,errcode);
       return 1;
   }
   else {
       ncols=nrows;
   }

   //Strong scaling
   //Weak scaling would set nrl=N and ncl=M
   if (N%nrows==0 && M%ncols==0) {
       nrl = N/nrows;
       ncl = M/ncols;
   }
   else {
       cout<<"Number of ranks should divide the total number of rows/cols evenly.\n";
       MPI_Abort(MPI_COMM_WORLD,errcode);
   }

    double **u=new double*[nrl+2];
    double **w=new double*[nrl+2];
    double **diffs=new double*[nrl+2];

    double *uptr=new double[(nrl+2)*(ncl+2)];
    double *wptr=new double[(nrl+2)*(ncl+2)];
    double *dptr=new double[(nrl+2)*(ncl+2)];

    for (int i=0;i<nrl+2;++i,uptr+=ncl+2) {
       u[i] = uptr;
    }
    for (int i=0;i<nrl+2;++i,wptr+=ncl+2) {
       w[i] = wptr;
    }
    for (int i=0;i<nrl+2;++i,dptr+=ncl+2) {
       diffs[i] = dptr;
    }

    //create Cartesian topology
    int ndims=2;
    int dims[2]={nrows,ncols};
    int periods[2]={0,0};
    int reorder=0;
    MPI_Comm grid_comm;
    int grid_rank;
    int grid_coords[2];

    MPI_Cart_create(MPI_COMM_WORLD, ndims, dims, periods, reorder, &grid_comm);
    MPI_Comm_rank(grid_comm, &grid_rank);
    MPI_Cart_coords(grid_comm, grid_rank, ndims, grid_coords);

    int direction=0;
    int displ=1;
    MPI_Cart_shift(grid_comm, direction, displ, &up, &down);

    direction=1;
    MPI_Cart_shift(grid_comm, direction, displ, &left, &right);

    //Set up the MPI type for columns
    MPI_Datatype col;
    MPI_Type_vector(nrl+2,1,ncl+2,MPI_DOUBLE,&col);
    MPI_Type_commit(&col);

    int nrequests=8;
    MPI_Request requests[nrequests];

    //Initialize solution matrices
    for ( int i = 0; i <= nrl+1; i++ ) {
         for (int j = 0; j <= ncl+1; j++ ) {
             u[i][j] = 0.;
             w[i][j] = 0.;
         }
    }
    set_bcs(nrl, ncl, nrows, ncols, grid_coords, u);

    diffInterval=1;

    double time0=MPI_Wtime();

  // Compute steady-state solution
  // The diffs array is mostly to make it simpler to extend the testing interval

  if (rank==0) {
      cout<<"Running until the difference is <="<<epsilon<<" with size "<<N<<"x"<<M<<"\n";
  }

  while ( iterations<=MAX_ITER ) {
     // reset diff each time through to get max abs diff
     // max_element doesn't work great for twod arrays and is often slow
     diff=.8*epsilon;

     MPI_Irecv(&u[0][1], ncl, MPI_DOUBLE, up, tag, MPI_COMM_WORLD, &requests[0]);
     MPI_Irecv(&u[nrl+1][1], ncl, MPI_DOUBLE, down, tag, MPI_COMM_WORLD, &requests[1]);
     MPI_Irecv(&u[0][0], 1, col, left, tag, MPI_COMM_WORLD, &requests[2]);
     MPI_Irecv(&u[0][ncl+1], 1, col, right, tag, MPI_COMM_WORLD, &requests[3]);

     MPI_Isend(&u[1][1], ncl, MPI_DOUBLE, up, tag, MPI_COMM_WORLD, &requests[4]);

     MPI_Isend(&u[nrl][1],ncl, MPI_DOUBLE, down, tag, MPI_COMM_WORLD, &requests[5]);
     MPI_Isend(&u[0][ncl], 1, col, right, tag, MPI_COMM_WORLD, &requests[6]);
     MPI_Isend(&u[0][1], 1, col, left, tag, MPI_COMM_WORLD, &requests[7]);

     MPI_Waitall(nrequests,requests,MPI_STATUS_IGNORE);

     for (int i=1; i<=nrl;i++) {
        for (int j=1;j<=ncl;j++) {
            w[i][j] = (u[i-1][j] + u[i+1][j] + u[i][j-1] + u[i][j+1])/4.0;
            diffs[i][j] = abs(w[i][j] - u[i][j]);
         }
     }
     if (iterations%diffInterval==0) {
        for (int i=1; i<=nrl;i++) {
           for (int j=1;j<=ncl;j++) {
           if (diff<diffs[i][j]) {
           diff=diffs[i][j];
           }
       }
     }
         //Find max of diff in all the processors.
         MPI_Allreduce(&diff,&gdiff,1,MPI_DOUBLE,MPI_MAX,MPI_COMM_WORLD);
         if (gdiff <= epsilon)
             break;
     }

     //Set halo values
     /*
     */
     for (int j=0; j<=ncl+1; j++) {
         w[0][j]=u[0][j];
         w[nrl+1][j]=u[nrl+1][j];
     }
     for (int i=0;i<=nrl+1; i++) {
         w[i][0]=u[i][0];
         w[i][ncl+1]=u[i][ncl+1];
     }

     for (int i=0; i<=nrl+1;i++) {
        for (int j=0;j<=ncl+1;j++) {
            u[i][j] = w[i][j];
        }
     }

     set_bcs(nrl, ncl, nrows, ncols, grid_coords, u);

     iterations++;
  } //end of computation 

  if (iterations>MAX_ITER) {
     if (rank==0) {
        cout<<"Warning: maximum iterations exceeded\n";
     }
  }

  double totalTime=(MPI_Wtime()-time0);
  if (rank==0) {
      cout << "Completed "<<iterations<<" iterations; time "<<totalTime<<endl;
  }

  // Write solution to output file
  ofstream fp;
  string fname=argv[2]+to_string(grid_coords[0])+to_string(grid_coords[1]);
  fp.open(fname,ios::out);
  for (int i = 1; i <= nrl; i++) {
     for (int j = 1; j <= ncl; j++) {
        fp<<u[i][j]<<" ";
     }
     fp<<"\n";
  }
  fp.close();

  //Type_free not really necessary but good practice
   MPI_Type_free(&col);
   MPI_Comm_free(&grid_comm);
   MPI_Finalize();

   return 0;

}

void set_bcs(int nr, int nc, int nrows, int ncols, int* coords, double **u) {

  /* Set boundary values.
   * This has an ice bath on the top edge.
   * Note: when plotting, 0,0 is the bottom.
   */

  double topBC=0.;
  double bottomBC=100.;
  double leftBC=100.;
  double rightBC=100.;

  if (coords[0]==0) {
     for (int i=0;i<=nc+1;++i){
         u[0][i]=topBC;
     }
  } else if (coords[0]==nrows-1) {
      for (int i=0;i<=nc+1;++i){
         u[nr+1][i]=bottomBC;
      }
  }
  if (coords[1]==0) {
      for (int i=0;i<=nr+1;++i){
         u[i][0]=leftBC;
      }
  } else if (coords[1]==ncols-1) {
     for (int i=0;i<=nr+1;++i){
        u[i][nc+1]=rightBC;
     }
  }

}
